arangod/Aql/IResearchViewExecutor.tpp

```diff
@@ -130,51 +130,101 @@ lookupCollection(arangodb::transaction::Methods& trx, DataSourceId cid,
+          return cmp.ascending ? lhs_stored->score < rhs_stored->score
+                               : lhs_stored->score > rhs_stored->score;
+        }
+      } else {
+        auto res = basics::VelocyPackHelper::compare(lhs_stored->slice,
+                                                     rhs_stored->slice, true);
+        if (res != 0) {
+          return (kSortMultiplier[size_t{cmp.ascending}] * res) < 0;
+        }
       }
+      ++lhs_stored;
+      ++rhs_stored;
     }
     return false;
   }
 
-  bool compareInput(size_t lhsIdx, float_t const* rhs_scores) const noexcept {
-    auto lhs_scores = &_scoreBuffer[lhsIdx * _numScoreRegisters];
-    for (auto const& cmp : _scoresSort) {
-      if (lhs_scores[cmp.first] != rhs_scores[cmp.first]) {
-        return cmp.second ? lhs_scores[cmp.first] < rhs_scores[cmp.first]
-                          : lhs_scores[cmp.first] > rhs_scores[cmp.first];
+  template<typename StoredValueProvider>
+  bool compareInput(size_t lhsIdx, float_t const* rhs_scores,
+                    StoredValueProvider const& stored) const noexcept {
+    TRI_ASSERT(_heapSortValues.size() > lhsIdx * _heapSort.size());
+    auto lhs_values = &_heapSortValues[lhsIdx * _heapSort.size()];
+    for (auto const& cmp : _heapSort) {
+      if (cmp.isScore()) {
+        if (lhs_values->score != rhs_scores[cmp.source]) {
+          return cmp.ascending ? lhs_values->score < rhs_scores[cmp.source]
+                               : lhs_values->score > rhs_scores[cmp.source];
+        }
+      } else {
+        auto value = stored(cmp);
+        auto res =
+            basics::VelocyPackHelper::compare(lhs_values->slice, value, true);
+        if (res != 0) {
+          return (kSortMultiplier[size_t{cmp.ascending}] * res) < 0;
+        }
       }
+      ++lhs_values;
     }
     return false;
   }
 
 #ifdef ARANGODB_ENABLE_MAINTAINER_MODE
   bool descScore() const noexcept {
-    return !_scoresSort.empty() && !_scoresSort.front().second;
+    return !_heapSort.empty() && _heapSort.front().isScore() &&
+           !_heapSort.front().ascending;
   }
 #endif
 
  private:
-  size_t _numScoreRegisters;
-  std::span<std::pair<size_t, bool> const> _scoresSort;
-  std::span<float_t const> _scoreBuffer;
+  std::span<HeapSortValue const> _heapSortValues;
+  std::span<HeapSortElement const> _heapSort;
 };
 
+velocypack::Slice getStoredValue(irs::bytes_view storedValue,
+                                 HeapSortElement const& sort) {
+  TRI_ASSERT(!sort.isScore());
+  [[maybe_unused]] auto const totalSize = storedValue.size();
+  VPackSlice slice{storedValue.data()};
+  if (slice.isNull()) {
+    return slice;
+  }
+  size_t size = 0;
+  size_t i = 0;
+  while (i < sort.fieldNumber) {
+    size += slice.byteSize();
+    TRI_ASSERT(size <= totalSize);
+    slice = VPackSlice{slice.end()};
+    ++i;
+  }
+  TRI_ASSERT(!slice.isNone());
+  if (sort.postfix.empty()) {
+    return slice;
+  } else {
+    return slice.get(sort.postfix);
+  }
```

Comment:
```suggestion
  auto* start = storedValue.data();
  [[maybe_unused]] auto* end = start + storedValue.size();
  velocypack::Slice slice{start};
  if (slice.isNull()) {
    return slice;
  }
  for (size_t i = 0; i != sort.fieldNumber; ++i) {
    start += slice.byteSize();
    TRI_ASSERT(start < end);
    slice = VPackSlice{start};
  }
  TRI_ASSERT(!slice.isNone());
  if (sort.postfix.empty()) {
    return slice;
  }
  TRI_ASSERT(!slice.isObject());
  return slice.get(sort.postfix);
```

Reply:
I suggest this because it's add more useful asserts and doesn't read byte size twice